### PR TITLE
JDK-8322042: HeapDumper should perform merge on the current thread instead of VMThread

### PR DESCRIPTION
**Pull request by alexmenkov**
The fix updated HeapDumper to always perform merge on the current thread.

Testing: tier1-5, all HeapDump-related tests
  Covered heap dumping scenarios:
    - `jcmd GC.heap_dump` command;
    - `HotSpotDiagnosticMXBean.dumpHeap()`;
    - `HeapDumpBeforeFullGC`, `HeapDumpAfterFullGC` VM options;
    - `HeapDumpOnOutOfMemoryError` VM option.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8322042](https://bugs.openjdk.org/browse/JDK-8322042): HeapDumper should perform merge on the current thread instead of VMThread (**Enhancement** - P4)

### Reviewers
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to [e373d793](https://git.openjdk.org/jdk/pull/18571/files/e373d793646a0753b94c3a28a41af0733300c228)
 * [Kevin Walls](https://openjdk.org/census#kevinw) (@kevinjwalls - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18571/head:pull/18571` \
`$ git checkout pull/18571`

Update a local copy of the PR: \
`$ git checkout pull/18571` \
`$ git pull https://git.openjdk.org/jdk.git pull/18571/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18571`

View PR using the GUI difftool: \
`$ git pr show -t 18571`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18571.diff">https://git.openjdk.org/jdk/pull/18571.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18571#issuecomment-2030861876)

### CODE REVIEW

**Comment by alexmenkov** [comment]
added serviceability label

**Review by y1yang0 — COMMENTED** [review]
- jcmd GC.heap_dump command; `AttachListenerThread`
- HotSpotDiagnosticMXBean.dumpHeap(); `JavaThread`
- HeapDumpBeforeFullGC, HeapDumpAfterFullGC VM options; `VMThread`
- HeapDumpOnOutOfMemoryError VM option. `VMThread`
Mabye we can always use AttachListenerThread(via Handshake) or new virtual thread?

File: src/hotspot/share/services/heapDumper.cpp [hunk 1]
```diff
@@ -2659,17 +2645,10 @@ int HeapDumper::dump(const char* path, outputStream* out, int compression, bool
...
-    // perform heapdump file merge operation in the current thread prevents us
-    // from occupying the VM Thread, which in turn affects the occurrence of
-    // GC and other VM operations.
-    merger.do_merge();
-  } else {
-    // otherwise, performs it by VM thread
-    VM_HeapDumpMerge op(&merger);
-    VMThread::execute(&op);
-  }
+  // perform heapdump file merge operation in the current thread prevents us
```
Comment (sspitsyn) [inline src/hotspot/share/services/heapDumper.cpp:2648]: Nit: Better to start with a capital letter.

**Review by sspitsyn — APPROVED** [review]

**Comment by sspitsyn** [comment]
This looks good and simpler to me.

**Comment by alexmenkov** [comment]
> * jcmd GC.heap_dump command; `AttachListenerThread`
> 
>     * HotSpotDiagnosticMXBean.dumpHeap(); `JavaThread`
> 
>     * HeapDumpBeforeFullGC, HeapDumpAfterFullGC VM options; `VMThread`
> 
>     * HeapDumpOnOutOfMemoryError VM option. `VMThread`
>       Mabye we can always use AttachListenerThread(via Handshake) or new virtual thread?

AttachListenerThread may not exist (it's created on demand).
Heap dump for HeapDumpOnOutOfMemoryError is performed on the current java thread.
I'm not sure I follow the proposal. What if the profit in using separate thread for merging?

File: src/hotspot/share/services/heapDumper.cpp (hunk 1 above)
Comment (alexmenkov) [reply src/hotspot/share/services/heapDumper.cpp]: Fixed.

**Comment by alexmenkov** [comment]
> Are we saying there is never any need to perform the merge in a VM Operation? Originally (JDK-8306441) it's either done in the attach thread, or a VM operation if we are in another thread. But maybe that was just being cautious.

Correct.
Originally it was implemented for AttachListener thread only for safety (and heap dump using jcmd was point of interest).
I didn't find issues with concurrent dumping (actually with concurrent merging, dumping itself is performed in VM_HeapDumper at safepoint).

**Comment by kevinjwalls** [comment]
Thanks Alex, happy you've looked into that. 8-)

**Review by kevinjwalls — APPROVED** [review]

**Pull request closed without merging**